### PR TITLE
Add artifact functions in ethers `getContractFactoryFromArtifact` and `getContractAtFromArtifact`

### PR DESCRIPTION
**Pull request by kanej**
In `hardhat-ethers`, allow passing of artifacts (resolved via custom code) into an equivalent of the `getContractFactory` call, this is instead of resolving by contract name.

The intention here is to support custom contract artifact resolution but still allow the leveraging of `hardhat-ethers`.

Two new functions have been added to support this without altering the existing signature:

* getContractFactoryFromArtifact
* getContractAtFromArtifact

Relates to #1716

## Example Usage

```typescript
import { ethers } from "hardhat";
import { Artifacts } from "hardhat/internal/artifacts";
import { Artifact } from "hardhat/types";

export async function getArtifact(contract: string): Promise<Artifact> {
  const artifactsPath: string = "./artifacts";

  const artifacts = new Artifacts(artifactsPath);
  return artifacts.readArtifact(contract);
}

async function main() {
  const artifact = await getArtifact("Greeter");

  const Factory = await ethers.getContractFactoryFromArtifact(artifact);

  const greeter = await Factory.deploy("Hello, Hardhat!");

  await greeter.deployed();

  console.log("Greeter deployed to:", greeter.address);

  const response = await greeter.greet();

  console.log("First Response", response);

  const alternateGreeter = await ethers.getContractAtFromArtifact(
    artifact,
    greeter.address
  );

  const secondResponse = await alternateGreeter.greet();

  console.log("Second Response", secondResponse);
}

// We recommend this pattern to be able to use async/await everywhere
// and properly handle errors.
main().catch((error) => {
  console.error(error);
  process.exitCode = 1;
});
```

## TODO

* [x] support artifact in signature of `getContractFactoryFromArtifact`
* [x] support artifact in signature of `getContractAtFromArtifact`
* [x] resolve type issue in consuming lib
* [x] determine and update any docs affected

### CODE REVIEW

File: .changeset/smooth-worms-end.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+"@nomiclabs/hardhat-ethers": patch
```
Comment (kanej) [inline .changeset/smooth-worms-end.md:2]: Given that this is changing the type signatures (and so can break existing code) should this be `minor`?

Comment (kanej) [reply .changeset/smooth-worms-end.md:2]: We have opted to add functions rather alter existing signatures. Hence this can be reduced to minor update.

**Comment by kanej** [comment]
I have test locally with newly created `advanced typescript project`, and we no longer get the type clash with typechain overriding the method signatures - getContractFactoryFromArtifact, getContractAtFromArtifact

File: packages/hardhat-ethers/README.md [hunk 2]
```diff
@@ -60,9 +60,12 @@ function getContractFactory(name: string, signer: ethers.Signer): Promise<ethers
 
 function getContractFactory(name: string, factoryOptions: FactoryOptions): Promise<ethers.ContractFactory>;
 
+function getContractFactoryFromArtifact(artifact: Artifact, signer?: ethers.Signer): Promise<ethers.ContractFactory>;
```
[the remark this answers is not in the document]
Comment (kanej) [reply packages/hardhat-ethers/README.md]: Good catch, it should be both. I have added in all the combinations of signatures for the current ethers interface (we had left out some functions/signature combinations) - so it now consistent with how the readme show the `getContractFactory` function.

**Review by fvictorio — APPROVED** [review]

**Pull request merged**